File: backend/models/oncall.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Boolean, Text, JSON, Time
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from database import Base


class OnCallSchedule(Base):
    __tablename__ = "oncall_schedules"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    org_id = Column(Integer, ForeignKey("organizations.id", ondelete="CASCADE"), nullable=True)
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    timezone = Column(String(100), default="UTC")
    rotation_type = Column(String(50), default="weekly")  # daily, weekly, custom
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    rotations = relationship("OnCallRotation", back_populates="schedule", cascade="all, delete-orphan")
    overrides = relationship("OnCallOverride", back_populates="schedule", cascade="all, delete-orphan")


class OnCallRotation(Base):
    __tablename__ = "oncall_rotations"

    id = Column(Integer, primary_key=True, index=True)
    schedule_id = Column(Integer, ForeignKey("oncall_schedules.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    order_index = Column(Integer, default=0)
    start_date = Column(DateTime(timezone=True), nullable=True)

    schedule = relationship("OnCallSchedule", back_populates="rotations")
    user = relationship("User")


class OnCallOverride(Base):
    __tablename__ = "oncall_overrides"

    id = Column(Integer, primary_key=True, index=True)
    schedule_id = Column(Integer, ForeignKey("oncall_schedules.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    start_time = Column(DateTime(timezone=True), nullable=False)
    end_time = Column(DateTime(timezone=True), nullable=False)
    reason = Column(String(255), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    schedule = relationship("OnCallSchedule", back_populates="overrides")
    user = relationship("User")


class EscalationPolicy(Base):
    __tablename__ = "escalation_policies"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    org_id = Column(Integer, ForeignKey("organizations.id", ondelete="CASCADE"), nullable=True)
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    repeat_count = Column(Integer, default=3)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    steps = relationship("EscalationStep", back_populates="policy", cascade="all, delete-orphan", order_by="EscalationStep.step_order")


class EscalationStep(Base):
    __tablename__ = "escalation_steps"

    id = Column(Integer, primary_key=True, index=True)
    policy_id = Column(Integer, ForeignKey("escalation_policies.id", ondelete="CASCADE"), nullable=False)
    step_order = Column(Integer, nullable=False)
    delay_minutes = Column(Integer, default=5)
    notify_type = Column(String(50), default="user")  # user, schedule, webhook
    notify_target_id = Column(Integer, nullable=True)
    notify_via = Column(JSON, default=list)  # ["email", "sms", "slack"]

    policy = relationship("EscalationPolicy", back_populates="steps")

File: frontend/src/api/oncall.ts
import client from './client'

export interface OnCallSchedule {
  id: number
  name: string
  description?: string
  timezone: string
  rotation_type: string
  is_active: boolean
  created_at: string
}

export interface OnCallRotation {
  id: number
  schedule_id: number
  user_id: number
  order_index: number
  user_name?: string
  user_email?: string
}

export interface OnCallOverride {
  id: number
  schedule_id: number
  user_id: number
  start_time: string
  end_time: string
  reason?: string
  user_name?: string
}

export interface EscalationPolicy {
  id: number
  name: string
  description?: string
  repeat_count: number
  is_active: boolean
  created_at: string
}

export interface EscalationStep {
  id: number
  policy_id: number
  step_order: number
  delay_minutes: number
  notify_type: string
  notify_target_id?: number
  notify_via: string[]
}

export const oncallApi = {
  listSchedules: () => client.get<OnCallSchedule[]>('/oncall/schedules'),
  createSchedule: (data: Partial<OnCallSchedule>) => client.post<OnCallSchedule>('/oncall/schedules', data),
  deleteSchedule: (id: number) => client.delete(`/oncall/schedules/${id}`),

  listRotations: (scheduleId: number) => client.get<OnCallRotation[]>(`/oncall/schedules/${scheduleId}/rotations`),
  addRotation: (scheduleId: number, data: Partial<OnCallRotation>) => client.post(`/oncall/schedules/${scheduleId}/rotations`, data),
  deleteRotation: (rotationId: number) => client.delete(`/oncall/rotations/${rotationId}`),

  listOverrides: (scheduleId: number) => client.get<OnCallOverride[]>(`/oncall/schedules/${scheduleId}/overrides`),
  addOverride: (scheduleId: number, data: Partial<OnCallOverride>) => client.post(`/oncall/schedules/${scheduleId}/overrides`, data),

  listEscalations: () => client.get<EscalationPolicy[]>('/oncall/escalations'),
  createEscalation: (data: Partial<EscalationPolicy>) => client.post<EscalationPolicy>('/oncall/escalations', data),
  deleteEscalation: (id: number) => client.delete(`/oncall/escalations/${id}`),

  listSteps: (policyId: number) => client.get<EscalationStep[]>(`/oncall/escalations/${policyId}/steps`),
  addStep: (policyId: number, data: Partial<EscalationStep>) => client.post(`/oncall/escalations/${policyId}/steps`, data),
}
